File: next/kmp/core/src/commonMain/kotlin/org/dweb_browser/core/ipc/IpcPool.kt
package org.dweb_browser.core.ipc

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineName
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import org.dweb_browser.core.help.types.IMicroModuleManifest
import org.dweb_browser.helper.Debugger
import org.dweb_browser.helper.SuspendOnce1
import org.dweb_browser.helper.UUID
import org.dweb_browser.helper.ioAsyncExceptionHandler
import org.dweb_browser.helper.randomUUID
import org.dweb_browser.pure.http.PureStream

val debugIpcPool = Debugger("ipcPool")

val kotlinIpcPool = IpcPool()

/**
 * IpcPool跟上下文对应，跟body流对应
 * Context(Kotlin,Worker,Front)
 * */
open class IpcPool {
  companion object {
    private fun randomPoolId() = "kotlin-${randomUUID()}"
  }

  val scope =
    CoroutineScope(CoroutineName("ipc-pool-kotlin") + ioAsyncExceptionHandler + SupervisorJob())

  /**每一个ipcPool都会绑定一个body流池,只有当不在同一个IpcPool的时候才需要互相拉取*/
  val poolId: UUID = randomPoolId()

  override fun toString() = "IpcPool@poolId=$poolId<uid=$poolId>"

  /**
   * 所有的ipc对象实例集合
   */
  private val ipcSet = mutableSetOf<Ipc>()

  /**
   * 所有的委托进来的流的实例集合
   */
  private val streamPool = mutableMapOf<String, PureStream>()
  suspend fun createIpc(
    endpoint: IpcEndpoint,
    pid: Int,
    locale: IMicroModuleManifest,
    remote: IMicroModuleManifest,
    autoStart: Boolean = false,
    startReason: String? = null,
  ) = Ipc(
    pid = pid,
    endpoint = endpoint,
    locale = locale,
    remote = remote,
    pool = this,
  ).also { ipc ->
    safeCreatedIpc(ipc, autoStart, startReason)
  }

  internal suspend fun safeCreatedIpc(
    ipc: Ipc,
    autoStart: Boolean,
    startReason: String?,
  ) {
    /// 保存ipc，并且根据它的生命周期做自动删除
    debugIpcPool("createIpc", ipc)
    ipcSet.add(ipc)
    /// 自动启动
    if (autoStart) {
      scope.launch {
        ipc.start(reason = startReason ?: "autoStart")
      }
    }
    ipc.onClosed {
      ipcSet.remove(ipc)
      debugIpcPool("removeIpc", ipc)
    }
  }

  /**关闭信号*/
  suspend fun awaitDestroyed() = runCatching {
    scope.coroutineContext[Job]!!.join();
    null
  }.getOrElse { it }

  val isDestroyed get() = scope.coroutineContext[Job]!!.isCancelled

  val destroy = SuspendOnce1 { cause: CancellationException ->
    val oldSet = ipcSet.toSet()
    ipcSet.clear()
    oldSet.forEach { ipc ->
      ipc.close()
    }
    scope.cancel(cause)
  }

}

File: desktop-dev/src/core/ipc/IpcPool.ts
import { createSignal } from "../../helper/createSignal.ts";
import { Ipc } from "../index.ts";

let ipc_pool_uid_acc = 0;

/**每一个worker 都会创建单独的IpcPool */
export class IpcPool {
  constructor(readonly poolName: string) {
    this.poolId = this.poolName + this.poolId;
  }

  readonly poolId = `-worker-${ipc_pool_uid_acc++}`;

  /**
   * 所有的ipc对象实例集合
   */
  #ipcSet = new Set<Ipc>();
  /**
   * 所有的委托进来的流的实例集合
   */
  #streamPool = new Map<String, ReadableStream>();

  /**安全的创建ipc */
  safeCreatedIpc(ipc: Ipc, autoStart: boolean, startReason?: string) {
    /// 保存ipc，并且根据它的生命周期做自动删除
    this.#ipcSet.add(ipc);
    // 自动启动
    if (autoStart) {
      ipc.start(startReason ?? "autoStart");
    }
    ipc.onClosed(() => {
      this.#ipcSet.delete(ipc);
      console.log("ipcpool-remote-ipc", ipc);
    });
  }

  //#region  close
  #destroySignal = createSignal();
  onDestory = this.#destroySignal.listen;
  async destroy() {
    this.#destroySignal.emit();
    for (const _ipc of this.#ipcSet) {
      await _ipc.close();
    }
    this.#ipcSet.clear();
  }
  // close end
}

// 这是一个简单的hashCode实现，用于计算字符串的hashCode
export function hashString(s: string): number {
  let hash = 0;
  for (let i = 0; i < s.length; i++) {
    // 使用charCodeAt获取字符的Unicode值，这个值在0-65535之间
    const charCode = s.charCodeAt(i);
    // 使用了一种称为“旋转哈希”的技术，通过将上一个哈希值左旋然后加上新字符的哈希值来生成新的哈希值
    hash = (hash << 5) - hash + charCode;
    // 使用按位异或运算符将hash值限制在一个32位的整数范围内
    hash = hash & hash;
  }
  return hash;
}

export const workerIpcPool = new IpcPool("desktop");
